### PR TITLE
restricted multipliers

### PR DESCRIPTION
**Pull request by olaugh**

### CODE REVIEW

**Comment by olaugh** [comment]
```
   A B C D E F G H I J K L M N O      Player 1                 EEGIPRW   38
   ------------------------------  -> Player 2                 EELOSST   42
 1|=     '       =       '     = | --Tracking-----------------------------------
 2|  -       "       "       -   | AAAAAAAABBCCDDDDEEEEEEEFFGGHHIIIIIIIKLLLMMNNNNNNOOOOOOOPRRRRRSSTTTTTUUVWXYY??  77
 3|    -       '   '       -     |
 4|'     -       '       -     ' |
 5|        -           -         |
 6|  "       "       "       "   |
 7|    '     Q '   J       '     |
 8|=     '   U V A E     '     = |
 9|    '     I '   U       '     |
10|  "       Z       "       "   |
11|        -           -         |
12|'     -       '       -     ' |
13|    -       '   '       -     |
14|  -       "       "       -   |
15|=     '       =       '     = |
   ------------------------------

highest_possible_equity was invalid. value: 83.000000
Better play has been found than should have been possible for this anchor, highest_possible_equity was invalid. Move: 6A TOELESS 86 has equity 86.000000
Assertion failed: (!better_play_has_been_found(gen, anchor_highest_possible_equity)), function generate_moves, file move_gen.c, line 1258.
```

**Comment by olaugh** [comment]
> ```
>    A B C D E F G H I J K L M N O      Player 1                 EEGIPRW   38
>    ------------------------------  -> Player 2                 EELOSST   42
>  1|=     '       =       '     = | --Tracking-----------------------------------
>  2|  -       "       "       -   | AAAAAAAABBCCDDDDEEEEEEEFFGGHHIIIIIIIKLLLMMNNNNNNOOOOOOOPRRRRRSSTTTTTUUVWXYY??  77
>  3|    -       '   '       -     |
>  4|'     -       '       -     ' |
>  5|        -           -         |
>  6|  "       "       "       "   |
>  7|    '     Q '   J       '     |
>  8|=     '   U V A E     '     = |
>  9|    '     I '   U       '     |
> 10|  "       Z       "       "   |
> 11|        -           -         |
> 12|'     -       '       -     ' |
> 13|    -       '   '       -     |
> 14|  -       "       "       -   |
> 15|=     '       =       '     = |
>    ------------------------------
> 
> highest_possible_equity was invalid. value: 83.000000
> Better play has been found than should have been possible for this anchor, highest_possible_equity was invalid. Move: 6A TOELESS 86 has equity 86.000000
> Assertion failed: (!better_play_has_been_found(gen, anchor_highest_possible_equity)), function generate_moves, file move_gen.c, line 1258.
> ```

haven't debugged at all but seems like it's short the perpendicular 3 points for the S on TLS, check out shadow_start_nonplaythrough

**Comment by olaugh** [comment]
Fixed the TOELESS bug, here's the next failure I get in autoplay
```
   A B C D E F G H I J K L M N O   -> Player 1                 AEIIRST   299
   ------------------------------     Player 2                 AAEEILW   312
 1|=     '       =       F A V A | --Tracking-----------------------------------
 2|  -       "       "   E   -   | AEEEEEFHIIJLMNNNNNOOORTTUY?  27
 3|    -       '   '     U G     |
 4|'     -       '       D O   ' |
 5|        -           A   U     |
 6|  "       "     B " R   R "   |
 7|  A B S O L V E R   T   D     |
 8|=     '       L I P I D e   = |
 9|    '       '   G   S   S     |
10|  "       "       " T     "   |
11|        -       R   I T       |
12|'     -       Q I   C H E Z ' |
13|    -       ' U M     A X E D |
14|  -       "   O   "   W   K O |
15|=     '       P O N G Y     C |
   ------------------------------

highest_possible_equity was invalid. value: 68.000000
Better play has been found than should have been possible for this anchor, highest_possible_equity was invalid. Move: D1 SATIRI(S)E 70 has equity 70.000000
Assertion failed: (!better_play_has_been_found(gen, anchor_highest_possible_equity)), function generate_moves, file move_gen.c, line 1268.
```

**Comment by olaugh** [comment]
> Fixed the TOELESS bug, here's the next failure I get in autoplay
> 
> ```
>    A B C D E F G H I J K L M N O   -> Player 1                 AEIIRST   299
>    ------------------------------     Player 2                 AAEEILW   312
>  1|=     '       =       F A V A | --Tracking-----------------------------------
>  2|  -       "       "   E   -   | AEEEEEFHIIJLMNNNNNOOORTTUY?  27
>  3|    -       '   '     U G     |
>  4|'     -       '       D O   ' |
>  5|        -           A   U     |
>  6|  "       "     B " R   R "   |
>  7|  A B S O L V E R   T   D     |
>  8|=     '       L I P I D e   = |
>  9|    '       '   G   S   S     |
> 10|  "       "       " T     "   |
> 11|        -       R   I T       |
> 12|'     -       Q I   C H E Z ' |
> 13|    -       ' U M     A X E D |
> 14|  -       "   O   "   W   K O |
> 15|=     '       P O N G Y     C |
>    ------------------------------
> 
> highest_possible_equity was invalid. value: 68.000000
> Better play has been found than should have been possible for this anchor, highest_possible_equity was invalid. Move: D1 SATIRI(S)E 70 has equity 70.000000
> Assertion failed: (!better_play_has_been_found(gen, anchor_highest_possible_equity)), function generate_moves, file move_gen.c, line 1268.
> ```

playthrough anchor this time

**Comment by olaugh** [comment]
``` 
  A B C D E F G H I J K L M N O   -> Player 1                 FHN       443
   ------------------------------     Player 2                 EPQ       436
 1|=     '       W       '     = | --Tracking-----------------------------------
 2|  -       " B A N D O l E E R |   0
 3|    -       ' N '       T   E |
 4|'     -       Z       O U   A |
 5|        P I X E S   - G I E D |
 6|  "       "       "   H   T I |
 7|    '       C O R     A ' A E |
 8|=     '   R E L A T U M   E S |
 9|    '       L I M O     ' R   |
10|  "     B "   D I C O T   I   |
11|  J     U       F   -     O R |
12|' I   - G     K I T U L     O |
13|  V -   O W ' N E       V A R |
14|  E     N Y   I D "       Y A |
15|A S S i G N A T       '   E L |
   ------------------------------

highest_possible_equity was invalid. value: 0.000000
Better play has been found than should have been possible for this anchor, highest_possible_equity was invalid. Move: N1 F(E)H 15 has equity 3.000000
Assertion failed: (!better_play_has_been_found(gen, anchor_highest_possible_equity)), function generate_moves, file move_gen.c, line 1265.
```

**Comment by olaugh** [comment]
Bounds are currently valid, but probably looser than I intend them to be, even within the shadow_right function where I expect an improvement. Overall seeing a 2% slowdown. shadow is slower without any speedup in recursive_gen. I'll add the leftward restrictions and then update the shadow tests to see if the good bounds I got in #176 are still happening

**Comment by olaugh** [comment]
```
   A B C D E F G H I J K L M N O      Player 1                 AAEEEII   266
   ------------------------------  -> Player 2                 ILRSTTU   195
 1|=     '       W       '     T | --Tracking-----------------------------------
 2|  -       "   O N "       - O | AAABBCDDDEEEGHIIIIJKLMMNNOOOOPPRRTVWYY  38
 3|    -       ' X U       -   U |
 4|' Q U R S H   ' N     -     Z |
 5|        C   F A c I A       I |
 6|  "     E "   E L O N G A T E |
 7|    '   L   '   E       '   R |
 8|=     F E V E R S     '     = |
 9|    '   r   '   '       '     |
10|  "     A "       "       "   |
11|        T           -         |
12|D O N G S     '       -     ' |
13|    -       '   '       -     |
14|  -       "       "       -   |
15|=     '       =       '     = |
   ------------------------------

highest_possible_equity was invalid. value: 58.000000
Better play has been found than should have been possible for this anchor, highest_possible_equity was invalid. Move: 10A ULTR(A)IST 62 has equity 62.000000
Assertion failed: (!better_play_has_been_found(gen, anchor_highest_possible_equity)), function generate_moves, file move_gen.c, line 1263.
```

**Comment by olaugh** [comment]
Bounds appear to be valid again, played 20k games without a failing assert.
With leftward hooks (how many of those are there, anyway? can anchors be defined such that they don't exist?) restricted, but still not using leftx/rightx to restrict tile positions, we are overall 1.5% slower.

recursive_gen is finally slightly faster, but we're not breaking even yet. i will use leftx/rightx for restriction and update the shadow tests to make sure there aren't bugs making the bounds worse (worse is valid, so the assert in move_gen.c can't pick that up) in some cases. the extra time in shadow isn't bad, maybe 4 µs per movegen. but maybe it can be sped up too, i've barely tried

**Comment by olaugh** [comment]
I've done what I think I need to do use restrict tile placement using leftx and rightx but this is still 1% slower.
Autoplay and the newly added shadow tests look fine but I need to uncomment+update the old shadow tests to see if we get all the pruning we expect

File: src/impl/move_gen.c
Comment (olaugh) [inline src/impl/move_gen.c:1]: This is obviously better but I'm very surprised it made such a difference.
Overall speedup now, ~1.8%

File: test/tsrc/shadow_test.c [hunk 1]
```diff
@@ -509,6 +551,68 @@ void test_shadow_score() {
...
+      "1RAVIoLI2G3Q/1EXON1IN2E1P1A/1C1L3GEM2AHI/BEMUD2SHUTTlED/1D1E8AI1/"
+      "YET9IS1/ODA9ST1/W1JOLLER7 BII/EO 477/388 0 lex CSW21";
+  load_and_generate(game, move_list, player, shuttled_ravioli, "BII");
+
+  // M3 B(U)I 25
+  assert(
+      within_epsilon(anchor_get_highest_possible_equity(anchor_list, 0), 25));
+
+  // 9B (EXON)B(IN) 19
+  // This should actually be 15 for (EXON)I(IN). The I being in the rightx set
```
Comment (olaugh) [inline test/tsrc/shadow_test.c:566]: why is this not the 15 for (EXON)I(IN) now?

Comment (olaugh) [reply test/tsrc/shadow_test.c]: Fixed now

**Comment by olaugh** [comment]
2% speedup

**Review by jvc56 — APPROVED** [review]
LGTM. Appreciate the additional tests!

**Pull request merged**